### PR TITLE
Add List API Endpoint for `RemoteRepository` and `RemoteOrganization`

### PR DESCRIPTION
**Pull request by saadmk11**
closes https://github.com/readthedocs/readthedocs.org/issues/7509

### CODE REVIEW

**Review by humitos — COMMENTED** [review]
This is a good start. However, I'm not sure if we have to implement this endpoint before https://github.com/readthedocs/readthedocs.org/issues/2759 (and https://github.com/readthedocs/readthedocs.org/issues/7169) because we are changing how `RemoteRepository` will behave completely.

Maybe the API response won't change, but would be good to double check that, tho. Want to take a look at this?

Also, we may need to be able to filter by user account (github, gitlab, etc) as we are currently doing on API v2:

- https://readthedocs.org/api/v2/remote/repo/?org=162925 (filter by organization)
- https://readthedocs.org/api/v2/remote/repo/?own=gitlab (filter by account)

(you can see your own API URLs with the `org=` and `own=` values at https://readthedocs.org/dashboard/import/)

The idea of using `/remote` as a namespace may be good as well. I'd call it `/remotes/repositories/` in the new APIv3.

In APIv2 we also have `/remote/account` and `/remote/org` that I think they will be required here as well (could be done in a different PR, but should be considered in the whole implementation).

If you have the time and energy to work on these things, I'm happy to help you with the process, chat and/or talk about it.

**Comment by saadmk11** [comment]
> This is a good start. However, I'm not sure if we have to implement this endpoint before #2759 (and #7169) because we are changing how `RemoteRepository` will behave completely.
> 
> Maybe the API response won't change, but would be good to double check that, tho. Want to take a look at this?

Sure! 👍  I did not see the changes being done with remote repositories. I'll look into the changes and see if there is anything we need to change about this API.

> Also, we may need to be able to filter by user account (github, gitlab, etc) as we are currently doing on API v2:

I was thinking about that, but did not know what are the requirements for API V3 are. Thanks for letting me know 💯 


> The idea of using `/remote` as a namespace may be good as well. I'd call it `/remotes/repositories/` in the new APIv3.

This sound like a good idea. 👍 

> In APIv2 we also have `/remote/account` and `/remote/org` that I think they will be required here as well (could be done in a different PR, but should be considered in the whole implementation).

Okay, so we need all the API endpoints that are required for the import process.

> If you have the time and energy to work on these things, I'm happy to help you with the process, chat and/or talk about it.
Sure, I would love to work on this, recently working with different API's and liking it a lot, and RTD has one of the best designed API 💯 

It would be great if you are able to help me with the process, Let me know when you are free to talk or chat about it, maybe on slack?

**Comment by humitos** [comment]
> It would be great if you are able to help me with the process, Let me know when you are free to talk or chat about it, maybe on slack?

I'm on Slack on Europe time. Feel free to ping me there when you have time.

**Comment by saadmk11** [comment]
Marking this as blocked as me and @humitos talked, this should be implemented after #2759 (and #7169)

**Comment by humitos** [comment]
@saadmk11 this could be a good pull request to cleanup and finish now we have the new structure deployed.

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1684]: I don't thinks this query string is required. I suppose that we will end up with an endpoint `/api/v3/remote/organizations/<slug>/` instead.

However, this `name=` attribute could be used as the `__contains` filter.

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1631]: These organizations are not "importable" :smile:

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1732]: I think that by default we shouldn't return the `organization` object in the response. We can make this attribute expandable (`?expand=organization`).

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1763]: I think we can remove this one. Is there a good use case?

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1765]: I think we should use the remote organization `slug`, instead. Most of the API is based/using `slug` now.

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1762]: Same here. I think we should use this as `__contains`. If this is the use case, we should explain it better.

We could also use `name__contains=` as attribute name as well. How are we calling these attributes in other APIv3 endpoints? We should follow the same pattern.

File: readthedocs/api/v3/serializers.py
Comment (humitos) [inline readthedocs/api/v3/serializers.py:920]: I think we can make `project` field expandable as well (`?expand=project`).

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1743]:
```suggestion
                    "clone_url": "https://github.com/organization/project.git",
```

Make sure the example matches the organization. There are some URLs pointing to `rtd` but the organization object says `organization`.

File: readthedocs/api/v3/serializers.py
Comment (humitos) [inline readthedocs/api/v3/serializers.py:934]: I think we are missing some fields here:

* `private`
* `default_branch`

File: readthedocs/api/v3/tests/test_remoteorganizations.py
Comment (humitos) [inline readthedocs/api/v3/tests/test_remoteorganizations.py:38]: Would be good to have a test that uses the `name=` filter which is the one we will use the most.

File: readthedocs/api/v3/tests/test_remoterepositories.py
Comment (humitos) [inline readthedocs/api/v3/tests/test_remoterepositories.py:64]: Same here. Would be good to have a test that uses `?name=` filter.

File: readthedocs/api/v3/views.py
Comment (humitos) [inline readthedocs/api/v3/views.py:465]: I'd create a `RemoteQuerySetMixin` in `api/v3/mixins.py` with this line, and use it in both views. We are doing that for the other views.

`RemoteRepositoryViewSet` may need to overwrite it tho, to add the `.annotate`, but that's fine.

File: readthedocs/api/v3/views.py
Comment (humitos) [inline readthedocs/api/v3/views.py:454]: `.select_related` will be only required if `?expand=organization` is passed.

**Review by humitos — CHANGES_REQUESTED** [review]
Looks great! I left some comment with a few changes I'd like to see before merging this PR.

I'm also pinging @agjohnson here who is the one using this endpoint for the search-as-you-type in the new theme.

File: docs/api/v3.rst
Comment (saadmk11) [reply docs/api/v3.rst:1763]: I don't think so, `vcs_provider` should be enough, but I saw it in the API v2, so I added it as I did not know the specific requirements for the API v3. :)

File: docs/api/v3.rst
Comment (saadmk11) [reply docs/api/v3.rst:1762]: We are not using anything like `__contains`, for example:

https://github.com/readthedocs/readthedocs.org/blob/901100594bf6549eda973be84b1ab7ad44379d64/readthedocs/api/v3/filters.py#L20


https://github.com/readthedocs/readthedocs.org/blob/901100594bf6549eda973be84b1ab7ad44379d64/docs/api/v3.rst#L480

Comment (humitos) [reply docs/api/v3.rst]: Gotcha!

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1684]: Oh, sorry. If we are already using `name=` to filter by `icontains`, we should use the same here. No need to change it to `__contains`. I'm sorry for the confusion here.

File: docs/api/v3.rst
Comment (humitos) [inline docs/api/v3.rst:1810]:
```suggestion
    :query string organization: return remote repositories for specific remote organization (using remote organization ``slug``)
```

I think that should be lower case :)

File: readthedocs/api/v3/views.py
Comment (humitos) [inline readthedocs/api/v3/views.py:470]: Nice! I wasn't aware of the `is_expanded` function!

**Review by humitos — COMMENTED** [review]
This is great!

We only need to revert the `name__contains` :angel:

**Review by humitos — APPROVED** [review]
:tada:

**Comment by humitos** [comment]
I'm blocking this PR only on @agjohnson 's review.

**Comment by agjohnson** [comment]
I don't have any particular feedback here, the PR looks great. If there is anything that I need changed it can be updated later.

**Comment by humitos** [comment]
Thanks @saadmk11 for all your work here! I'm merging this PR.

**Pull request merged**
